File: packages/assets/iwi.go
package assets

import (
	"encoding/binary"
	"fmt"
	"io"
	"os"
	"sort"
)

type (
	IWI struct {
		Header  iwiHeader
		Info    iwiInfo
		Data    []byte
		RawData []byte
	}

	iwiHeader struct {
		Magic   [3]byte
		Version byte
	}

	iwiInfo struct {
		Format byte
		Usage  byte
		Width  uint16
		Height uint16
		Depth  uint16
	}

	iwiMipmapOffsets []int32
	iwiMipmap        struct {
		offset int32
		size   int32
	}
	iwiMipmaps []iwiMipmap
)

const (
	// IWI Versions
	IWI_VERSION_COD2 = 0x05 // CoD2
	IWI_VERSION_COD4 = 0x06 // CoD4
	IWI_VERSION_COD5 = 0x06 // CoD5

	// IWI Format
	IWI_FORMAT_ARGB32 = 0x01 // ARGB32
	IWI_FORMAT_RGB24  = 0x02 // RGB24
	IWI_FORMAT_GA16   = 0x03 // GA16
	IWI_FORMAT_A8     = 0x04 // A8
	IWI_FORMAT_DXT1   = 0x0B // DXT1
	IWI_FORMAT_DXT3   = 0x0C // DXT3
	IWI_FORMAT_DXT5   = 0x0D // DXT5
)

// mipMaps calculate mipmap offsets and sizes from the given offsets
func (ofs iwiMipmapOffsets) mipMaps(first int32, size int32) iwiMipmaps {
	m := make(iwiMipmaps, len(ofs))

	for i := 0; i < len(ofs); i++ {
		switch i {
		case 0:
			m[i] = iwiMipmap{
				offset: ofs[i],
				size:   size - ofs[i],
			}

		case len(ofs) - 1:
			m[i] = iwiMipmap{
				offset: first,
				size:   ofs[i] - first,
			}

		default:
			m[i] = iwiMipmap{
				offset: ofs[i],
				size:   ofs[i-1] - ofs[i],
			}
		}
	}

	return m
}

func (m iwiMipmaps) Len() int           { return len(m) }                // Sort interface - Len
func (m iwiMipmaps) Less(i, j int) bool { return m[i].size < m[j].size } // Sort interface-  Less
func (m iwiMipmaps) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }      // Sort interface - Swap

// Load
func (iwi *IWI) Load(filepath string) error {
	f, err := os.Open(filepath)
	if err != nil {
		return errorLogAndReturn(err)
	}
	defer f.Close()

	err = iwi.readHeader(f)
	if err != nil {
		return errorLogAndReturn(err)
	}

	err = iwi.readInfo(f)
	if err != nil {
		return errorLogAndReturn(err)
	}

	ofs := make(iwiMipmapOffsets, 4)
	err = binary.Read(f, binary.LittleEndian, &ofs)
	if err != nil {
		return errorLogAndReturn(err)
	}

	curr, err := f.Seek(0, io.SeekCurrent)
	if err != nil {
		return errorLogAndReturn(err)
	}

	stat, err := f.Stat()
	if err != nil {
		return errorLogAndReturn(err)
	}

	size := stat.Size()
	mms := ofs.mipMaps(int32(curr), int32(size))
	sort.Sort(sort.Reverse(mms))
	mm := mms[0]

	_, err = f.Seek(int64(mm.offset), io.SeekStart)
	if err != nil {
		return errorLogAndReturn(err)
	}

	data := make([]byte, mm.size)
	err = binary.Read(f, binary.LittleEndian, &data)
	if err != nil {
		return errorLogAndReturn(err)
	}

	iwi.Data = data
	err = iwi.decodeData()
	if err != nil {
		return errorLogAndReturn(err)
	}
	return nil
}

// readHeader
func (iwi *IWI) readHeader(f *os.File) error {
	err := binary.Read(f, binary.LittleEndian, &iwi.Header)
	if err != nil {
		return errorLogAndReturn(err)
	}

	if iwi.Header.Magic != [3]byte{'I', 'W', 'i'} {
		return fmt.Errorf("invalid magic: %s", string(iwi.Header.Magic[:]))
	}

	supportedVersions := []uint8{
		IWI_VERSION_COD2,
		IWI_VERSION_COD4,
		IWI_VERSION_COD5,
	}

	supported := false
	for _, sv := range supportedVersions {
		if sv == iwi.Header.Version {
			supported = true
			break
		}
	}

	if !supported {
		return fmt.Errorf("invalid IWi version: %v", iwi.Header.Version)
	}

	return nil
}

// readInfo
func (iwi *IWI) readInfo(f *os.File) error {
	err := binary.Read(f, binary.LittleEndian, &iwi.Info)
	if err != nil {
		return errorLogAndReturn(err)
	}
	return nil
}

// decodeData
func (iwi *IWI) decodeData() error {
	var err error
	switch iwi.Info.Format {
	case IWI_FORMAT_DXT1:
		iwi.RawData, err = decodeDXT1(iwi.Data, uint(iwi.Info.Width), uint(iwi.Info.Height))
		if err != nil {
			return errorLogAndReturn(err)
		}
		return nil
	case IWI_FORMAT_DXT3:
		iwi.RawData, err = decodeDXT3(iwi.Data, uint(iwi.Info.Width), uint(iwi.Info.Height))
		if err != nil {
			return errorLogAndReturn(err)
		}
		return nil
	case IWI_FORMAT_DXT5:
		iwi.RawData, err = decodeDXT5(iwi.Data, uint(iwi.Info.Width), uint(iwi.Info.Height))
		if err != nil {
			return errorLogAndReturn(err)
		}
		return nil
	default:
		return fmt.Errorf("unsupported decode format: %v", iwi.Info.Format)
	}
}

File: src/assets/iwi.rs
use crate::utils::{
    binary, decode::decode_dxt1, decode::decode_dxt3, decode::decode_dxt5, error::Error, Result,
};
use std::{
    fs::File,
    io::{Seek, SeekFrom},
    path::PathBuf,
    str,
};
use valid_enum::ValidEnum;

pub struct IWi {
    width: u16,
    height: u16,
    data: Vec<u8>,
}

struct IWiHeader {
    magic: [u8; 3],
    version: u8,
}

#[derive(Clone, Copy)]
struct IWiInfo {
    format: u8,
    usage: u8,
    width: u16,
    height: u16,
    depth: u16,
}

#[derive(Clone, Copy)]
struct IWiMipMap {
    offset: i32,
    size: i32,
}

#[derive(ValidEnum)]
#[valid_enum(u8)]
pub enum IWiVersion {
    V5 = 0x05,
    V6 = 0x06,
}

#[derive(ValidEnum)]
#[valid_enum(u8)]
pub enum IWiFormat {
    ARGB32 = 0x01,
    RGB24 = 0x02,
    GA16 = 0x03,
    A8 = 0x04,
    DXT1 = 0x0B,
    DXT3 = 0x0C,
    DXT5 = 0x0D,
}

impl IWi {
    pub fn load(file_path: PathBuf) -> Result<IWi> {
        let mut file = File::open(&file_path)?;
        let header = Self::read_header(&mut file)?;
        let info = Self::read_info(&mut file)?;

        let offsets = binary::read_vec::<i32>(&mut file, 4)?;
        let current_offset = binary::current_offset(&mut file)?;
        let file_size = file.seek(SeekFrom::End(0))?;
        let mipmap = Self::calculate_highest_mipmap(offsets, current_offset, file_size);
        let raw_texture_data = binary::read_vec::<u8>(&mut file, mipmap.size as usize)?;
        if raw_texture_data.len() == 0 {
            return Err(Error::new(String::from("texture data length is 0")));
        }

        let data = Self::decode_data(raw_texture_data, header.version, info)?;

        Ok(IWi {
            width: info.width,
            height: info.height,
            data,
        })
    }

    fn read_header(file: &mut File) -> Result<IWiHeader> {
        let magic = binary::read_vec::<u8>(file, 4)?;
        if magic != [b'I', b'W', b'i'] {
            return Err(Error::new(format!(
                "invalid magic: {}",
                str::from_utf8(&magic).unwrap()
            )));
        }

        let version = binary::read::<u8>(file)?;
        match IWiVersion::valid(version) {
            Some(_) => (),
            None => return Err(Error::new(format!("invalid IWi version {}", version))),
        }

        Ok(IWiHeader {
            magic: magic.try_into().unwrap(),
            version,
        })
    }

    fn read_info(file: &mut File) -> Result<IWiInfo> {
        let format = binary::read::<u8>(file)?;
        let usage = binary::read::<u8>(file)?;
        let width = binary::read::<u16>(file)?;
        let height = binary::read::<u16>(file)?;
        let depth = binary::read::<u16>(file)?;

        Ok(IWiInfo {
            format,
            usage,
            width,
            height,
            depth,
        })
    }

    fn calculate_highest_mipmap(offsets: Vec<i32>, first: u64, size: u64) -> IWiMipMap {
        let mut mipmaps: Vec<IWiMipMap> = Vec::new();

        let offsets_len = offsets.len();
        for i in 0..offsets_len {
            if i == 0 {
                mipmaps.push(IWiMipMap {
                    offset: offsets[i],
                    size: size as i32 - offsets[i],
                });
            } else if i == (offsets_len - 1) {
                mipmaps.push(IWiMipMap {
                    offset: first as i32,
                    size: offsets[i] - first as i32,
                });
            } else {
                mipmaps.push(IWiMipMap {
                    offset: offsets[i],
                    size: offsets[i - 1] - offsets[i],
                })
            }
        }

        let mut max_idx = 0;
        for i in 0..mipmaps.len() {
            if mipmaps[i].size > mipmaps[max_idx].size {
                max_idx = i;
            }
        }

        return mipmaps[max_idx];
    }

    fn decode_data(data: Vec<u8>, version: u8, info: IWiInfo) -> Result<Vec<u8>> {
        match IWiFormat::valid(info.format) {
            Some(IWiFormat::DXT1) => Ok(decode_dxt1(data, info.width, info.height)),
            Some(IWiFormat::DXT3) => Ok(decode_dxt3(data, info.width, info.height)),
            Some(IWiFormat::DXT5) => Ok(decode_dxt5(data, info.width, info.height)),
            _ => Err(Error::new(format!(
                "unsupported decode format {}",
                info.format
            ))),
        }
    }
}
